include/RAJA/pattern/tensor/internal/ET/MultiplyOperator.hpp

```diff
@@ -0,0 +1,973 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   RAJA header defining expression template behavior for operator*
+ *
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-22, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_pattern_tensor_ET_MultiplyOperator_HPP
+#define RAJA_pattern_tensor_ET_MultiplyOperator_HPP
+
+
+
+namespace RAJA
+{
+namespace internal
+{
+namespace expt
+{
+  //forward
+  class TensorBlockConcreteBase;
+
+
+
+
+  namespace ET
+  {
+
+
+    /*!
+     * Provides default multiply, multiply add, and multiply subtract
+     * operations.
+     *
+     * If the operands are both matrices, we perform a matrix-matrix multiply.
+     * Otherwise, we perform element-wise operations.
+     */
+    template<typename LEFT_OPERAND_TYPE, typename RIGHT_OPERAND_TYPE, class ENABLE = void>
```

Comment:
I assume we have `class ENABLE = void` to play nice with the specializations below? If so how does the compiler define between the default and the specialized versions of MultiplyOperator.

Reply:
the ENABLE=void is for the std::enable_if's.  If none of the enable_if statements are valid, it will fall back on the default implementation.

The default is the element-wise multiplication, and all of the specializations define things like vec*mat, mat*vec, mat*mat, which all have specific linear algebra implementations (not element-wise)